tests/ui/redundant_type_annotations.rs

```diff
@@ -0,0 +1,151 @@
+#![allow(unused)]
+#![warn(clippy::redundant_type_annotations)]
+
+#[derive(Debug, Default)]
+struct Cake<T> {
+    _data: T,
+}
+
+fn make_something<T: Default>() -> T {
+    T::default()
+}
+
+fn make_cake<T: Default>() -> Cake<T> {
+    Cake::<T>::default()
+}
+
+fn plus_one<T: std::ops::Add<u8, Output = T>>(val: T) -> T {
+    val + 1
+}
+
+struct Pie {
+    inner: u32,
+}
+
+enum Pizza {
+    One,
+    Two,
+}
+
+fn return_a_string() -> String {
+    String::new()
+}
+
+fn return_a_struct() -> Pie {
+    Pie { inner: 5 }
+}
+
+fn return_an_enum() -> Pizza {
+    Pizza::One
+}
+
+fn return_an_int() -> u32 {
+    5
+}
+
+impl Pie {
+    fn return_an_int(&self) -> u32 {
+        self.inner
+    }
+
+    fn associated_return_an_int() -> u32 {
+        5
+    }
+
+    fn new() -> Self {
+        Self { inner: 5 }
+    }
+
+    fn associated_return_a_string() -> String {
+        String::from("")
+    }
+
+    fn test_method_call(&self) {
+        let v: u32 = self.return_an_int(); // Should lint
+    }
+}
+
+fn test_generics() {
+    // The type annotation is needed to determine T
+    let _c: Cake<i32> = make_something();
+
+    // The type annotation is needed to determine the topic
+    let _c: Cake<u8> = make_cake();
+
+    // This should lint (doesn't)
+    let _c: Cake<u8> = make_cake::<u8>();
+
+    // This should lint (doesn't)
+    let _c: u8 = make_something::<u8>();
+
+    // This should lint (doesn't)
+    let _c: u8 = plus_one(5_u8);
+
+    // Annotation needed otherwise T is i32
+    let _c: u8 = plus_one(5);
+}
+
+fn test_non_locals() {
+    // This shouldn't lint
+    fn _arg(x: u32) -> u32 {
+        x
+    }
+
+    // This could lint, but probably shouldn't
+    let _closure_arg = |x: u32| x;
+}
+
+fn test_complex_types() {
+    // Shouldn't lint, since the literal will be i32 otherwise
+    let _u8: u8 = 128;
+
+    // Should lint (doesn't)
+    let _tuple_i32: (i32, i32) = (12, 13);
+
+    // Shouldn't lint, since the tuple will be i32 otherwise
+    let _tuple_u32: (u32, u32) = (1, 2);
+
+    // Should lint, since the type is determined by the init value (doesn't)
+    let _tuple_u32: (u32, u32) = (3_u32, 4_u32);
+
+    // Should lint (doesn't)
+    let _array: [i32; 3] = [5, 6, 7];
+
+    // Shouldn't lint
+    let _array: [u32; 2] = [8, 9];
+}
+
+fn test_functions() {
+    // Everything here should lint
+
+    let _return: String = return_a_string();
+
+    let _return: Pie = return_a_struct();
+
+    let _return: Pizza = return_an_enum();
+
+    let _return: u32 = return_an_int();
+
+    let _return: String = String::new();
+
+    let new_pie: Pie = Pie::new();
+
+    let _return: u32 = new_pie.return_an_int();
+
+    let _return: String = String::from("test");
```

Comment:
This currently doesn't trigger a lint, even if the function comment says it should. I'm not quite sure why, as `String::new()` etc are covered.

Reply:
Wow, I missed that, thanks!

I digged into it a bit, it turns out it is related to generics.
When I do `cx.tcx.type_of(func_defid)` of the function, I get `EarlyBinder(fn(T) -> Self {<Self as std::convert::From<T>>::from})`

If I then execute `subst_identity` or `skip_binder` on this, I get `fn(T) -> Self {<Self as std::convert::From<T>>::from}`

Then, when I do `func_ty.fn_sig(cx.tcx).output()` where `func_ty` is the type I wrote previously, I get `Binder(Self, [])`, then I interpret the return type as `Self` instead of `String`.

Do you know what I could use in this case? I tried having a look at other lints but I was not able to find anything, thanks!

Reply:
cc @xFrednet :D

Reply:
Sorry, I missed your question earlier :sweat_smile:. I'm not totally sure about this one, but you might want to take a look at:
* [`type_dependent_def`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/typeck_results/struct.TypeckResults.html#method.type_dependent_def) to resolve the method target
* Maybe [qpath_res](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/typeck_results/struct.TypeckResults.html#method.qpath_res) to resolve `Self`?

If both of these don't work, I'd probably ask on [Zulip](https://rust-lang.zulipchat.com/#narrow/stream/257328-clippy) the others might have a good idea for this one :)

Reply:
> Sorry, I missed your question earlier 😅. I'm not totally sure about this one, but you might want to take a look at:
> 
> * [`type_dependent_def`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/typeck_results/struct.TypeckResults.html#method.type_dependent_def) to resolve the method target
> * Maybe [qpath_res](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/typeck_results/struct.TypeckResults.html#method.qpath_res) to resolve `Self`?
> 
> If both of these don't work, I'd probably ask on [Zulip](https://rust-lang.zulipchat.com/#narrow/stream/257328-clippy) the others might have a good idea for this one :)

I've tried the two methods you suggested without any luck. I've asked on Zulip, let's see.

Thanks!